kvbc/tools/migrations/v4migration_tool/src/migration_bookeeper.cpp

```diff
@@ -0,0 +1,262 @@
+// Concord
+//
+// Copyright (c) 2018-2022 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0
+// License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#include <pthread.h>
+#include <rocksdb/filter_policy.h>
+#include <rocksdb/table.h>
+
+#include "migration_bookeeper.hpp"
+
+using namespace boost::program_options;
+using namespace concord::storage::rocksdb;
+using namespace concord::util;
+
+int BookKeeper::migrate(int argc, char* argv[]) {
+  if (!parse_args(argc, argv)) {
+    return EXIT_FAILURE;
+  }
+
+  initialize_migration();
+
+  LOG_INFO(logger_, "Starting the migration");
+  if (config_["migrate-to-v4"].as<bool>()) {
+    migrate(cat_kvbc_->getAdapter(), v4_kvbc_->getAdapter());
+  } else {
+    migrate(v4_kvbc_->getAdapter(), cat_kvbc_->getAdapter());
+  }
+
+  LOG_INFO(logger_, "Successfully completed Migration, thanks for your time !!!");
+  return EXIT_SUCCESS;
+}
+
+void BookKeeper::initialize_migration() {
+  number_of_batches_allowed_in_cycle_ = config_["max-point-lookup-batches"].as<std::uint64_t>();
+  const auto read_only = true;
+  const auto migrate_to_v4 = config_["migrate-to-v4"].as<bool>();
+
+  auto rodb = NativeClient::newClient(
+      config_["input-rocksdb-path"].as<std::string>(), read_only, NativeClient::DefaultOptions{});
+  auto wodb = NativeClient::newClient(
+      config_["output-rocksdb-path"].as<std::string>(), !read_only, NativeClient::DefaultOptions{});
```

Comment:
do we want to assert that input-rocksdb-path != output-rocksdb-path ?

Reply:
@cloudnoize , thanks for the comment. I am not checking this, since the DB can be the same aswell. So input DB path and output DB path can be same.